Packer crash

<!--- Please keep this note for the community --->

#### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

<!--- Thank you for keeping this note for the community --->

When filing a bug, please include the following headings if possible. Any
example text in this template can be deleted.

#### Overview of the Issue
<details>
panic: value is not a collection

goroutine 1 [running]:
github.com/zclconf/go-cty/cty.Value.LengthInt({{{0x962bba8, 0xc00013c5a9}}, {0x7dfcd80, 0xc00103b370}})
	github.com/zclconf/go-cty@v1.10.0/cty/value_ops.go:1087 +0x1aa
github.com/zclconf/go-cty/cty.Value.AsValueMap({{{0x962bba8, 0xc00013c5a9}}, {0x7dfcd80, 0xc00103b370}})
	github.com/zclconf/go-cty@v1.10.0/cty/value_ops.go:1318 +0x71
github.com/hashicorp/packer/hcl2template.(*Parser).decodeProvisioner(0x0, 0xc0010301a0, 0x0)
	github.com/hashicorp/packer/hcl2template/types.build.provisioners.go:110 +0x8ec
github.com/hashicorp/packer/hcl2template.(*Parser).decodeBuildConfig(0xc0004e4060, 0xc0004e61a0, 0xc001022000)
	github.com/hashicorp/packer/hcl2template/types.build.go:168 +0x8bb
github.com/hashicorp/packer/hcl2template.(*Parser).parseConfig(0xc001022000, 0xc00103f650, 0xc001022000)
	github.com/hashicorp/packer/hcl2template/parser.go:373 +0x1eb
github.com/hashicorp/packer/hcl2template.(*PackerConfig).Initialize(0xc001022000, {0x0})
	github.com/hashicorp/packer/hcl2template/parser.go:326 +0x59a
github.com/hashicorp/packer/command.(*BuildCommand).RunContext(0xc000b68f60, {0x9629c60, 0xc0009a9a00}, 0xc00014e000)
	github.com/hashicorp/packer/command/build.go:153 +0x82
github.com/hashicorp/packer/command.(*BuildCommand).Run(0xc000b68f60, {0xc000116070, 0x3, 0x3})
	github.com/hashicorp/packer/command/build.go:40 +0xc5
github.com/mitchellh/cli.(*CLI).Run(0xc000328780)
	github.com/mitchellh/cli@v1.1.2/cli.go:262 +0x5f8
main.wrappedMain()
	github.com/hashicorp/packer/main.go:262 +0xaf2
main.realMain()
	github.com/hashicorp/packer/main.go:49 +0xf3
main.main()
	github.com/hashicorp/packer/main.go:35 +0x19
</details>

packer build -on-error="ask" -only="test.googlecompute.rocky8" testing/packer/linux/

!!!!!!!!!!!!!!!!!!!!!!!!!!! PACKER CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!

Packer crashed! This is always indicative of a bug within Packer.
A crash log has been placed at "crash.log" relative to your current
working directory. It would be immensely helpful if you could please
report the crash with Packer[1] so that we can fix this.

[1]: https://github.com/hashicorp/packer/issues

!!!!!!!!!!!!!!!!!!!!!!!!!!! PACKER CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!

#### Reproduction Steps

Added rocky linux source to existing packer build config

### Packer version

Packer v1.8.0

### Simplified Packer Template

<details>

```
build {
  name = "test"

  source "source.googlecompute.centos7" {
    image_name        = "dbk-centos7-base-testing-v${formatdate("YYYYMMDDhhmmss", timestamp())}"
    image_description = "NCR Digital Banking CentOS 7 Base Test Image"
    image_family      = "dbk-centos-7-base-testing"
    machine_type      = "n1-standard-2"
    image_labels = var.centos7_labels

    source_image_project_id = [var.project_id]

    region             = var.region
    zone               = var.zone
    network_project_id = var.network_project_id
    subnetwork         = var.subnetwork
    disk_size          = var.disk_size

    service_account_email = var.service_account_email
    account_file          = var.account_file
    communicator          = "ssh"
    ssh_username          = "packer"
    use_iap               = true
    use_internal_ip       = true
    use_os_login          = false
    omit_external_ip      = true
    skip_create_image     = var.skip_create_image

    scopes = ["https://www.googleapis.com/auth/cloud-platform"]

    metadata = {
      enable-guest-attributes = "TRUE",
      enable-osconfig         = "TRUE",
      enable-oslogin          = "FALSE",
    }
  }

  source "source.googlecompute.centos8" {
    image_name        = "dbk-centos8-base-testing-v${formatdate("YYYYMMDDhhmmss", timestamp())}"
    image_description = "NCR Digital Banking CentOS 8 Base Test Image"
    image_family      = "dbk-centos-8-base-testing"
    machine_type      = "n1-standard-2"
    image_labels = var.centos8_labels

    source_image_project_id = [var.centos_source_image_project_id]

    region             = var.region
    zone               = var.zone
    network_project_id = var.network_project_id
    subnetwork         = var.subnetwork
    disk_size          = var.disk_size

    service_account_email = var.service_account_email
    account_file          = var.account_file
    communicator          = "ssh"
    ssh_username          = "packer"
    use_iap               = true
    use_internal_ip       = true
    use_os_login          = false
    omit_external_ip      = true
    skip_create_image     = var.skip_create_image

    scopes = ["https://www.googleapis.com/auth/cloud-platform"]

    metadata = {
      enable-guest-attributes = "TRUE",
      enable-osconfig         = "TRUE",
      enable-oslogin          = "FALSE",
    }
  }

 source "source.googlecompute.rocky8" {
    image_name        = "dbk-rocky8-base-testing-v${formatdate("YYYYMMDDhhmmss", timestamp())}"
    image_description = "NCR Digital Banking Rocky 8 Base Test Image"
    image_family      = "dbk-rocky-8-base-testing"
    machine_type      = "n1-standard-2"
    image_labels      = var.rocky8_labels

    source_image_project_id = [var.rocky_source_image_project_id]

    region             = var.region
    zone               = var.zone
    network_project_id = var.network_project_id
    subnetwork         = var.subnetwork
    disk_size          = var.disk_size

    service_account_email = var.service_account_email
    account_file          = var.account_file
    communicator          = "ssh"
    ssh_username          = "packer"
    use_iap               = true
    use_internal_ip       = true
    use_os_login          = false
    omit_external_ip      = true
    skip_create_image     = var.skip_create_image

    scopes = ["https://www.googleapis.com/auth/cloud-platform"]

    metadata = {
      enable-guest-attributes = "TRUE",
      enable-osconfig         = "TRUE",
      enable-oslogin          = "FALSE",
    }
  } 

  source "source.googlecompute.redhat7" {
    image_name        = "dbk-rhel7-base-testing-v${formatdate("YYYYMMDDhhmmss", timestamp())}"
    image_description = "NCR Digital Banking RedHat 7 Base Test Image"
    image_family      = "dbk-redhat7-base-testing"
    machine_type      = "n1-standard-2"
    source_image_project_id = [var.project_id]

    image_labels       = var.rhel7_labels
    region             = var.region
    zone               = var.zone
    network_project_id = var.network_project_id
    subnetwork         = var.subnetwork
    disk_size          = var.disk_size

    service_account_email = var.service_account_email
    account_file          = var.account_file
    communicator          = "ssh"
    ssh_username          = "packer"
    use_iap               = true
    use_internal_ip       = true
    use_os_login          = false
    omit_external_ip      = true
    skip_create_image     = var.skip_create_image

    scopes = ["https://www.googleapis.com/auth/cloud-platform"]

    metadata = {
      enable-guest-attributes = "TRUE",
      enable-osconfig         = "TRUE",
      enable-oslogin          = "FALSE",
    }
  }

  provisioner "shell" {
    inline = [
      "sudo mkdir /ansible",
      "sudo chown -R packer:packer /ansible",
      "sudo yum update -y",
      "sudo yum -y install epel-release",
      "sudo yum install python3-pip -y",
      "sudo pip3 install --upgrade pyyaml==5.4.1",
      "sudo yum -y install ansible",
      "sudo install -d -o ${build.User} ${local.packer_work_dir} ${local.ansible_staging_dir}",
      "sudo systemctl stop firewalld.service",
      "sudo systemctl disable firewalld.service",
      "ansible-galaxy collection install community.general",
      "ansible-galaxy collection install google.cloud",
      "ansible-galaxy collection install ansible.posix",
    ]
    override = {
      centos8 = {
        inline = [
          "sudo yum -y install epel-release",
          "sudo yum install python38-pip -y",
          "sudo yum install python38-pyyaml -y",
          "sudo yum -y install ansible",
          "sudo install -d -o packer ${local.packer_work_dir} ${local.ansible_staging_dir}",
          "sudo systemctl stop firewalld.service",
          "sudo systemctl disable firewalld.service",
          "ansible-galaxy collection install community.general",
          "ansible-galaxy collection install google.cloud",
          "ansible-galaxy collection install ansible.posix",
        ]
      }
      redhat6 = {
        inline = [
          "sudo yum -y update",
          "sudo yum -y install epel-release",
          "sudo yum -y install ansible",
        ]
      }
    remote_folder = "/var/tmp"  
    }
  }

  provisioner "ansible-local" {
    extra_arguments = [
      "-vvvvv",
      "--extra-vars",
      "'INSTANCES=ungrouped group_name=SaaS-GCP-Prod-Monitored jfrog_username=${var.jfrog_username} jfrog_token=${var.jfrog_token}'",
    ]
    playbook_dir      = "./ansible"
    playbook_file     = "./ansible/playbook.yaml"
    staging_directory = local.ansible_staging_dir
  }

  provisioner "shell" {
    scripts = [
      "./scripts/aidefix.sh",
      "./scripts/cleanup.sh",
    ]
    remote_folder    = "/var/tmp"
    valid_exit_codes = [0, 100]
  }
  post-processor "checksum" {
    checksum_types = ["md5", "sha512"]
  }

  post-processor "manifest" {
    output     = "manifest.json"
    strip_path = true
  }
}
```
</details>

### Operating system and Environment details

MacOS 12.2.1

### Log Fragments and crash.log files

[crash.log](https://github.com/hashicorp/packer/files/9108646/crash.log)

--- a/hcl2template/fixtures/malformed_override.pkr.hcl
+++ b/hcl2template/fixtures/malformed_override.pkr.hcl
@@ -1,0 +1,4 @@
+provisioner "shell-local" {
+  inline = ["echo 'hi'"]
+  override = "hello"
+}

--- a/hcl2template/fixtures/malformed_override_innards.pkr.hcl
+++ b/hcl2template/fixtures/malformed_override_innards.pkr.hcl
@@ -1,0 +1,6 @@
+provisioner "shell-local" {
+  inline = ["echo 'hi'"]
+  override = {
+    test = "hello"
+  }
+}

--- a/hcl2template/fixtures/well_formed_provisioner.pkr.hcl
+++ b/hcl2template/fixtures/well_formed_provisioner.pkr.hcl
@@ -1,0 +1,9 @@
+provisioner "shell-local" {
+  inline = ["echo 'hi'"]
+  override = {
+    test = {
+      "hello" = "new value"
+    }
+  }
+}
+

--- a/hcl2template/types.build.provisioners.go
+++ b/hcl2template/types.build.provisioners.go
@@ -104,9 +104,28 @@ func (p *Parser) decodeProvisioner(block *hcl.Block, ectx *hcl.EvalContext) (*Pr
 	}
 
 	if !b.Override.IsNull() {
+		if !b.Override.Type().IsObjectType() {
+			return nil, append(diags, &hcl.Diagnostic{
+				Severity: hcl.DiagError,
+				Summary:  "provisioner's override block must be an HCL object",
+				Subject:  block.DefRange.Ptr(),
+			})
+		}
+
 		override := make(map[string]interface{})
 		for buildName, overrides := range b.Override.AsValueMap() {
 			buildOverrides := make(map[string]interface{})
+
+			if !overrides.Type().IsObjectType() {
+				return nil, append(diags, &hcl.Diagnostic{
+					Severity: hcl.DiagError,
+					Summary: fmt.Sprintf(
+						"provisioner's override.'%s' block must be an HCL object",
+						buildName),
+					Subject: block.DefRange.Ptr(),
+				})
+			}
+
 			for option, value := range overrides.AsValueMap() {
 				buildOverrides[option] = hcl2shim.ConfigValueFromHCL2(value)
 			}

--- a/hcl2template/types.build.provisioners_test.go
+++ b/hcl2template/types.build.provisioners_test.go
@@ -1,0 +1,82 @@
+package hcl2template
+
+import (
+	"testing"
+
+	"github.com/hashicorp/hcl/v2"
+)
+
+func TestPackerConfig_ParseProvisionerBlock(t *testing.T) {
+	tests := []struct {
+		name                 string
+		inputFile            string
+		expectError          bool
+		expectedErrorMessage string
+	}{
+		{
+			"success - provisioner is valid",
+			"fixtures/well_formed_provisioner.pkr.hcl",
+			false,
+			"",
+		},
+		{
+			"failure - provisioner override is malformed",
+			"fixtures/malformed_override.pkr.hcl",
+			true,
+			"provisioner's override block must be an HCL object",
+		},
+		{
+			"failure - provisioner override.test is malformed",
+			"fixtures/malformed_override_innards.pkr.hcl",
+			true,
+			"provisioner's override.'test' block must be an HCL object",
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			cfg := PackerConfig{parser: getBasicParser()}
+			f, diags := cfg.parser.ParseHCLFile(test.inputFile)
+			if diags.HasErrors() {
+				t.Errorf("failed to parse input file %s", test.inputFile)
+				for _, d := range diags {
+					t.Errorf("%s", d)
+				}
+				return
+			}
+			provBlock := f.OutermostBlockAtPos(hcl.Pos{
+				Line:   1,
+				Column: 1,
+				Byte:   0,
+			})
+			_, diags = cfg.parser.decodeProvisioner(provBlock, nil)
+
+			if !diags.HasErrors() {
+				if !test.expectError {
+					return
+				}
+
+				t.Fatalf("unexpected success")
+			}
+
+			if !test.expectError {
+				for _, d := range diags {
+					t.Errorf("%s", d)
+				}
+			}
+
+			gotExpectedErr := false
+			for _, d := range diags {
+				if d.Summary == test.expectedErrorMessage {
+					gotExpectedErr = true
+				}
+
+				t.Logf("got error (expected): '%s'", d.Summary)
+			}
+
+			if !gotExpectedErr {
+				t.Errorf("never got expected error: '%s'", test.expectedErrorMessage)
+			}
+		})
+	}
+}

